axlearn/common/optimizers.py

```diff
@@ -1993,3 +2008,93 @@ def _update2(u: Tensor, param: OptParam):
+            context.
+
+    This function returns a new `PartitionedGradientTransformation` that
+    1. Puts all states of the wrapped optimizer on `offload_dst` through the partition function
+       during state initialization in the trainer.
+    2. Copies the states to `offload_src` before `optimizer.update` is called.
+    3. Copies the updated states to `offload_dst` after `optimizer.update` is called.
+
+    The .update function of the returned `PartitionedGradientTransformation` must be called within
+    a jit function.
+
+    Example usage:
+    ```python
+    your_opt = adamw_optimizer(...)
+    offloaded_opt = offload_optimizer(your_opt)
+    ```
+
+    Only wrap the optimizer that you actually want to offload with this function to avoid
+    unneseccary overhead. This is usually the optimizer that occupies the most HBM. For example,
+    when you have chained optimizers:
+    ```python
+    # Recommended
+    chain([
+        some_preprocessing(...),
+        clip_by_global_norm(...),
+        offload_optimizer(adamw_decoupled_optimizer(...)),
+    ])
+    # Not recommended
+    offload_optimizer(chain([
+        some_preprocessing(...),
+        clip_by_global_norm(...),
+        adamw_decoupled_optimizer(...),
+    ]))
+    ```
+
+    When using `skip_and_clip_by_global_norm` with this offload optimizer, you must wrap the entire
+    `skip_and_clip_by_global_norm` inside. Do not wrap the inner of `skip_and_clip_by_global_norm`
+    or you will get errors. Correct example:
+    ```
+    offloaded_opt = offload_optimizer(skip_and_clip_by_global_norm(inner=adamw_optimizer(...)))
+    ```
+    The reason is that `skip_and_clip_by_global_norm` conditionally chooses the previous optimizer
+    state and the updated new optimizer state using `jnp.where`, which doesn't support tensors on
+    `pinned_host` memory space.
+    """
+    optimizer = maybe_instantiate(optimizer)
+    if offload_src is None or offload_dst is None:
+        raise ValueError(
+            "offload_src and offload_dst cannot be None when using optimizer offloading."
+        )
+
+    logging.info("Optimizer offloading enabled.")
+
+    def init_fn(params: NestedOptParam):
+        return optimizer.init(params)
+
+    def update_fn(updates: optax.Updates, state: optax.OptState, params: NestedOptParam):
+        # TransferToMemoryKind let us change the memory kind of tensors without specifying the full
+        # sharding (i.e. jax.sharding.NamedSharding). Although there's no documentation about it,
+        # it's specified in the API signature. Reference:
+        # https://github.com/jax-ml/jax/blob/21f8885a9e104b8828c9a8b721eed0c68b622691/jax/_src/api.py#L2220
+        state = jax.device_put(state, TransferToMemoryKind(offload_src))
+        updates, state = optimizer.update(updates, state, params)
+        state = jax.device_put(state, TransferToMemoryKind(offload_dst), donate=True)
```

Comment:
Do we need explicit `device_put` calls here? Is it enough to specify the partition spec with the right memory_kind?

Reply:
Haven't tested with_sharding_constraint, if you are refering to that. However, specifying full sharding requires us to store the sharding when the partition fn is called, which is not preferred by John (see internal comments)

Reply:
I'm not sure whether we need `with_sharding_constraint`, since jit/pjit should already apply shardings returned by `partition_fn` to the states. What happens without these `device_put` calls?

If necessary, we can always invoke `partition_fn` here to compute the sharding on the fly (instead of storing them) and apply `with_sharding_constraint`.